src/syntax/Scanner.js

```diff
@@ -1172,11 +1173,30 @@ function isAtEnd() {
 
 function next() {
   index++;
-  updateCurrentCharCode();
+  if (currentCodePoint >= 0x10000) {
+    index++;
+  }
+  updateCurrentCodePoint();
+}
+
+function updateCurrentCodePoint() {
+  currentCodePoint = currentCharCode = input.charCodeAt(index);
+  nextCharCode = input.charCodeAt(index + 1) | 0;
+  if (isLeadSurrogate(currentCodePoint) && isTrailSurrogate(nextCharCode)) {
```

Comment:
One simple optimization would probably be to to check if the currentCodePoint is ASCII.

```js
function updateCurrentCodePoint() {
  currentCodePoint = input.charCodeAt(index);
  if (currentCodePoint < 128) return;
  ...
}
```